[#1] Wordle Game

annnhs opened the pull request.

annnhs committed:
.gitignore update

annnhs committed:
test: add answer test

annnhs committed:
feat: add check answer feature

annnhs committed:
feat: add UI feature

annnhs committed:
feat: add wordle game

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/Answer.java:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        return words.toArray(new String[0]);
+    }
```
Stream, Reader 이런 객체를 사용하면 꼭 리소스를 해제하는 작업이 필요해요.
지금 사용한 `BufferedReader` 는 `close()` 메서드를 호출하면 매개변수로 받은 리소스도 같이 해제 해줍니다!

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/Answer.java:
```diff
@@ -0,0 +1,117 @@
+import java.io.InputStreamReader;
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
```
이런 객체들이 `Answer` 객체가 꼭 의존해야 하는지 의문이네요
객체 이름만 들었을 땐 '매일 변하는 정답 단어' 라는 도메인 인 것 같은데 날짜, 리소스 등 여러 의존이 생기면서 복잡도가 올라가고 있어요.

추가로 import 해줄 때 '*' 처럼 Wildcard import 는 지양해야 한다고 합니다!
낱개로 하나씩 해주는 Absolute import 하는 방식으로 연습하면 좋을 것 같아요.
IDE 에서도 설정에서 지원하는 부분이니 찾아보시면 좋을 것 같아요.

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/Answer.java:
```diff
@@ -0,0 +1,117 @@
+
+    public boolean isNotContainInWords(final String guess) {
+        return !words.contains(guess);
+    }
```
사용자가 입력한 답안을 검증하는 로직이 Answer 에 있어서 좀 어색한 것 같네요.
역할과 책임을 잘 구분하는 연습이 필요해요.

개인적으로 역할과 책임을 잘 나누는 방법은 일단 좋은 객체 명 짓기 에서 시작한다고 생각하는데요,
예를 들어 '카페' 라는 시스템을 만든다면 역할로는 매니저, 캐셔, 바리스타, 손님 등이 있을 것 같아요.
그 다음은 각 역할에 맞는 책임을 부여해야 해요.

매니저는 정산, 재고 관리, 인사 관리, 급여 관리
캐셔는 주문 접수, 결제, 바리스타에게 메뉴 전달
바리스타는 메뉴 제조, 손님에게 음료 전달
손님은 주문, 요금 지불, 음료 섭취

등 이 될 것 같은데요.

만약 이렇게 역할과 책임을 구분 지었다 하더라도 객체 덩치가 커지면 또 분리하는게 더 유리할거에요.

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/Answer.java:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        return new String(result);
+    }
```
메서드가 15줄이 넘으면 안된다는 요구사항을 만족하지 못했어요😭

djawnstj commented on src/test/java/com/djawnstj/wordle/domain/AnswerTest.java:
```diff
@@ -0,0 +1,115 @@
+
+class AnswerTest {
+
+    @Test
+    @DisplayName("(현재 날짜 - 2021년 6월 19일) 구하기")
+    public void getDiffOfDays() {
+        // Given
+        int baseYear = 2021;
+        int baseMonth = 6;
+        int baseDayOfMonth = 19;
+
+        LocalDate baseDate = LocalDate.of(baseYear, baseMonth, baseDayOfMonth);
+        LocalDate currentDate = LocalDate.now();
+
+        // When
+        long diff = ChronoUnit.DAYS.between(baseDate, currentDate);
+        if (diff > Integer.MAX_VALUE || diff < Integer.MIN_VALUE) {
+            throw new ArithmeticException("The difference in number of days from the base date exceeds the range of int type.");
+        }
+
+        // Then
+        int diffOfDays = (int) diff;
+    }
```
테스트에 검증부가 아무 역할을 하지 않아요

djawnstj commented on src/test/java/com/djawnstj/wordle/domain/AnswerTest.java:
```diff
@@ -0,0 +1,115 @@
+    void compareWords() {
+        // Given
+        String answer = "abcce";
+        String guess = "jeaca";
+        String resultString = "XYYGX";
+
+        // When
+        char[] result = new char[guess.length()];
+        boolean[] answerUsed = new boolean[answer.length()];
+
+        for (int i = 0; i < guess.length(); i++) {
+            if (guess.charAt(i) == answer.charAt(i)) {
+                result[i] = 'G';
+                answerUsed[i] = true;
+                continue;
+            }
+            result[i] = 'X';
+        }
+
+        Map<Character, Integer> frequencyMap = new HashMap<>();
+        for (int i = 0; i < answer.length(); i++) {
+            if (!answerUsed[i]) {
+                frequencyMap.put(answer.charAt(i), frequencyMap.getOrDefault(answer.charAt(i), 0) + 1);
+            }
+        }
+
+        for (int i = 0; i < guess.length(); i++) {
+            if (result[i] == 'G') {
+                continue;
+            }
+            char currentChar = guess.charAt(i);
+            if (frequencyMap.containsKey(currentChar) && frequencyMap.get(currentChar) > 0) {
+                result[i] = 'Y';
+                frequencyMap.put(currentChar, frequencyMap.get(currentChar) - 1);
+            }
+        }
+
+        // Then
+        assertThat(new String(result)).isEqualTo(resultString);
+    }
```
전체적으로 테스트가 프로덕션 코드를 검증하는게 아니라 하드코딩으로 테스트에 로직이 박혀있어요😭
테스트는 프로덕션 코드를 호출해서 검증하는 작업이 필요해요.

djawnstj commented on src/main/java/com/djawnstj/wordle/Wordle.java:
```diff
@@ -0,0 +1,65 @@
+        getInput();
+    }
+
+    private void getInput() {
+        renderer.showInput();
+        final String input = scanner.next();
+
+        if (input.length() != 5) {
+            renderer.showIsNot5Words();
+            getInput();
+            return;
+        }
+
+        if (answer.isNotContainInWords(input)) {
+            renderer.showNotInWords();
+            getInput();
+            return;
+        }
+
+        checkInput(input);
+    }
```
`getInput` 인데 입력값을 검증하고 있어요.

djawnstj commented on src/main/java/com/djawnstj/wordle/Wordle.java:
```diff
@@ -0,0 +1,65 @@
+
+    private void finishWordle(final String result) {
+        renderer.showGameOver(numOfTry, result);
+        if (!Objects.equals(result, "GGGGG")) {
```
"GGGGG" 는 마치 매직 넘버 같은 느낌이에요.
구현을 모르는 사람이 와서 보면 어떤 의미인지 코드를 들여다 봐야 파악할 수 있어서 안좋은 방법인것 같네요.
어떻게 개선할 수 있을까요?

djawnstj reviewed (changes requested):
호성님!
이제 지금 제출한 코드를 객체지향 코드로 전환해봅시다.
하면서 어려운 부분 있으면 말씀 주세요~

annnhs committed:
test: Refactoring test code

- rename AnswerTest to FeedbackGeneratorTest
- separate code to DateUtilTest, TextReaderTest

annnhs committed:
refactor: Separate roles

- separate code because Answer class has too many roles
- rename Answer to FeedbackGenerator
- refactoring Wordle class

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/WordDictionary.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final String[] words;
+
+    public WordDictionary() {
+        this.words = initWords();
+    }
```
생성자가 너무 큰 책임을 갖고 있어요~
이렇게 될 경우, 이 클래스를 이용하는 사람이 의도하지 않은 동작이 발생할 수 있기 때문에 초기화 로직은 따로 호출하게 열어두는게 좋아요.

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/TextReader.java:
```diff
@@ -0,0 +1,33 @@
+            words.add(line);
+        }
+
+        inputStream.close();
+        reader.close();
```
피드백 반영 👍🏻

다만 닫는 코드까지 오기 전 예외가 발생 된다면 리소스 해제가 영원히 되지 못할 거 같아요.

또 지금 `inputStream` 과 `reader` 를 닫아 줬는데 이렇게 따로 닫아줘야 할까요?
그렇다면 `BufferedReader` 의 생성자로 넘겨 준 `InputStreamReader(inputStream)` 은 안 닫아줘도 괜찮을까요?

djawnstj reviewed (changes requested):
호성님~
조금만 더 리팩토링 해봅시다!

annnhs committed:
refactor: Move out initWords() from constructor

annnhs committed:
refactor: Use try-with-resources for resource management of InputStreamReader and BufferedReader

annnhs committed:
refactor: Add Answer class to replace targetWord

annnhs committed:
refactor: Replace targetWord to Answer class

- extract method in finishWordle()
- update scanner.next() to nextLine()

annnhs committed:
refactor: Move to util package

- move Wordle class to domain package

annnhs committed:
refactor: Add logic to verify answer when generating Answer class

annnhs committed:
fix: Remove spaces in player's input

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/FeedbackGenerator.java:
```diff
@@ -0,0 +1,61 @@
+    final char EXIST = 'Y';
+    final char WRONG = 'X';
+
+    public String generateFeedback(final String guess, final Answer answer) {
+        final boolean[] answerUsed = new boolean[answer.length()];
+
+        final char[] greenFeedback = checkGreenFeedback(guess, answer, answerUsed);
+
+        final Map<Character, Integer> frequencyMap = createFrequencyMap(answerUsed, answer);
+
+        final char[] feedback = checkYellowFeedback(guess, greenFeedback, frequencyMap);
+
+        return new String(feedback);
+    }
```
`guess` 가 사용자가 입력한 단어인거처럼 보이는데요,
여전히 `FeedbackGenerator` 는 `guess` 가 게임에서 사용하기 적합하지 않은 단어도 전달받고 있네요😭

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/FeedbackGenerator.java:
```diff
@@ -0,0 +1,61 @@
+    private char[] checkGreenFeedback(final String guess, final Answer answer, final boolean[] answerUsed) {
+        final char[] feedback = new char[guess.length()];
+
+        for (int i = 0; i < guess.length(); i++) {
+            if (guess.charAt(i) == answer.charAt(i)) {
+                feedback[i] = CORRECT;
+                answerUsed[i] = true;
+                continue;
+            }
+            feedback[i] = WRONG;
+        }
```
for 문에서 두번째 인자로 들어가는 값은 순회를 할 때마다 호출하게 돼요.
그러면 `guess.length()` 가 매번 호출된다는 얘긴데 이렇게 매번 호출해서 검증하도록 하는 방법은 시간 복잡도 측면에서 좋지 못할 것 같네요!

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/Wordle.java:
```diff
@@ -0,0 +1,87 @@
+        numOfTry += 1;
+        final String feedback = feedbackGenerator.generateFeedback(input, answer);
+
+        if (answer.isEqualTo(input) || numOfTry == 6) {
+            finishWordle(feedback);
+            return;
+        }
```
'6' 이 어떤 의미를 가진 숫자일까요?
이 코드를 처음 보는 사람도 코드를 한번에 파악할 수 있도록 작성해 주셔야 해요.

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/Wordle.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    private boolean isWrongAnswer(final String feedback) {
+        final String correctWord = "GGGGG";
+        return !Objects.equals(feedback, correctWord);
+    }
```
"GGGGG" 도 어떤 의미인지 파악하려면 별도의 문맥이 필요한 것 같네요!

또 입력한 단어가 정답인지 아닌지 파악하는 책임이 `Wordle` 객체에 있다는게 적절할까요?
이 객체의 역할이 뭔지 조금 더 고민이 필요할 것 같습니다

djawnstj commented on src/main/java/com/djawnstj/wordle/ui/UIRenderer.java:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    private void addToResults(final String guess) {
+        StringBuilder output = new StringBuilder();
+        for (final char c : guess.toCharArray()) {
+            if (c == 'G') {
+                String GREEN = "\uD83D\uDFE9";
+                output.append(GREEN);
+            } else if (c == 'Y') {
+                String YELLOW = "\uD83D\uDFE8";
+                output.append(YELLOW);
+            } else if (c == 'X') {
+                String GREY = "⬜";
+                output.append(GREY);
+            }
+        }
+
+        results.add(output.toString());
+    }
```
게임 결과를 상태로 가지게 하고, 결과를 추가하는 로직을 `UIRenderer` 에 넣은 이유가 궁금합니다!

djawnstj commented on src/main/java/com/djawnstj/wordle/util/TextReader.java:
```diff
@@ -0,0 +1,35 @@
+
+public class TextReader {
+
+    public String[] convertFileToList(final String fileName) throws IOException {
+        List<String> words = new ArrayList<>();
+        ClassLoader classLoader = getClass().getClassLoader();
+
+        InputStream inputStream = classLoader.getResourceAsStream(fileName);
+        if (inputStream == null) {
+            throw new AssertionError("inputStream is null. Cannot find: " + fileName);
+        }
+
+        try (InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
+             BufferedReader reader = new BufferedReader(inputStreamReader)) {
+
+            String line;
+            while ((line = reader.readLine()) != null) {
+                words.add(line);
+            }
+        } finally {
+            inputStream.close();
+        }
+
+        return words.toArray(new String[0]);
+    }
```
메서드의 길이가 15줄을 넘어버렸네요

djawnstj reviewed (changes requested):
호성님 리뷰 남겼습니다~
그리고 전체적으로 테스트 코드가 부족해요.
모든 public 메서드에 대해선 예외케이스까지 테스트를 작성해야 합니다!

피드백 남기고 궁금한 점도 질문 남겼는데 확인 부탁드려요!

annnhs committed:
refactor: Refactor with Feedback, ValidWord class

- generate Feedback with Feedback class
- move and rename addToResults() in UIRenderer to makeFeedbackVisual() in FeedbackGenerator
- rename class Answer to ValidWord
- use ValidWord class to player's guess
- move isWrongAnswer() in Wordle to Feedback class

annnhs committed:
test: Add test code

- rename FeedbackGeneratorTest to FeedbackTest
- rename TextReaderTest to WordDictionaryTest
- add WordValidatorTest

annnhs committed:
refactor: Move feedbacks in Feedback to Wordle class

annnhs committed:
refactor: UIRenderer only has the role of rendering

djawnstj commented on src/main/java/com/djawnstj/wordle/domain/ValidWord.java:
```diff
@@ -1,30 +1,30 @@
     public boolean isEqualTo(final String guess) {
-        return answer.equals(guess);
+        return word.equals(guess);
     }
```
사용하지 않는 메서드는 없애주는게 좋아요!

djawnstj reviewed (commented):
호성님 수고하셨습니다~
코드가 충분히 많이 깔끔해진 것 같아요.
간단하지만 많은걸 얻어가는 프로젝트가 되었으면 합니다!

djawnstj reviewed (approved):
👍🏻

annnhs committed:
refactor: extract method in TextReader class

- remove unused method in ValidWord
